### ISSUE TITLE
findAreaPaths() taking way too long to calculate paths in unowned large continents

### ISSUE THREAD

**Issue by JohnMTorgerson** [JohnMTorgerson/Viking#4]
This was something we anticipated, but hadn't yet played enough games to have seen until now. It seems to happen a lot in the River Castles map towards the end of the game, when it tries to take over one of the large continents that it has no countries in, which of course yields a large number of possible paths. We've seen it take around 30 seconds or so, just at a guess. Maybe even longer once or twice.

We intend to try to replicate this scenario with some outputs that will tell us both the total number of paths calculated and the amount of time it took to find them all, but haven't done so yet.

Possible solutions:

1) See what we can do to further optimize findAreaPaths(); not hopeful this alone will work

2) Simply tell findAreaPaths() to stop looking after 100(?) or 1,000(?) or 10,000(?) paths, assuming that it would have found at least one that was good enough by then.

3) Explicitly tell findAreaPaths() to stop looking after it found its first exhaustive path that ends on a border (plus any other criteria we care about).

**Comment by JohnMTorgerson** [comment]
We discovered through testing that findAreaPaths was not responsible for the bulk of the time when finding paths. Rather, the prune checkPaths loop in pickBestTakeoverPaths() was the culprit. We tried a couple different optimizations and found that removing elements from the checkPaths ArrayList was the problem. So instead of removing the paths we want to discard from the checkPaths ArrayList, we're adding the paths we want to keep to a new ArrayList, and then overwriting checkPaths with the new list at the end. It's literally like 1,000 times faster. Previously, pruning check paths took more than 1 minute on large lists (~500,000-800,000 long), and now takes less than 100ms on the biggest lists we've seen.

The whole path-finding operation still takes 1-2 seconds on the very largest lists (on the order of several hundred thousand paths long), which is still too long, but good enough for now. We may revisit this again later on to cut it down some more, but I think a thousand-fold improvement is good enough for the moment.

**Comment by JohnMTorgerson** [comment]
Despite our previous optimization, tonight I ran into another instance of it taking an absurd amount of time to calculate paths, this time in Russia on the Europe Large 3 - Napoleonic Wars map.

![screen-shot-2015-11-05-at-4 48 13-am](https://cloud.githubusercontent.com/assets/6146526/10966930/83ef6914-837c-11e5-8dbf-16795cd7ef74.png)

I force quit Lux after waiting for ten full minutes. As you can see in the screenshot, the bot had not yet chatted the number of paths it had found, which is still turned on, which means it was taking all that time just to find the paths, never mind to pick one. Either that or Lux simply froze. Further testing on this map is warranted, and I'm reopening the issue. If it's possible to take >10 minutes on a continent of 31 countries, I shudder to think how long it might take if it ever ran across one that had, say 60 countries. It would probably be days. Clearly, we need to limit the number of paths it's searching for after all, or otherwise optimize findAreaPaths().

**Comment by JohnMTorgerson** [comment]
Resolved this issue by using a global variable (pathCount) to keep track of the number of terminal paths we've found, and when findAreaPaths() gets to 100,000 it just stops looking. Sometimes this might result in not finding a path that gets to every country in the area, but getAreaTakeoverPaths() is already equipped to handle that by calling findAreaPaths() again on whatever countries are leftover. Even if it has to recurse a bunch of times to get all of the countries, that's okay, because that process essentially scales linearly.

I haven't actually timed how long it takes to find 100,000 paths, but it seems to be on the order of 2-3 seconds. That's pretty long, but it's not totally ridiculous. It only happens on the largest continents, and then only when we don't already own very many countries in them (a situation that doesn't occur too often now that we have the landgrab objective), and when it does happen, it should only need to find paths once or twice per turn. So it's a noticeable delay, but I think it's acceptable. We can always cut it down to fewer than 100,000 or try to find further optimizations later if we want., but for now I'm closing this issue.
